pkg/gke/membership.go

```diff
@@ -0,0 +1,152 @@
+package gke
+
+import (
+	"context"
+	"fmt"
+
+	gkehub "cloud.google.com/go/gkehub/apiv1beta1"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"github.com/googleapis/gax-go"
+	gkehubpb "google.golang.org/genproto/googleapis/cloud/gkehub/v1beta1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	capg "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
+)
+
+const (
+	AuthorityIssuer = "https://kubernetes.default.svc.cluster.local"
+)
+
+type GKEMembershipClient interface {
+	CreateMembership(context context.Context, req *gkehubpb.CreateMembershipRequest, opts ...gax.CallOption) (*gkehub.CreateMembershipOperation, error)
+	GetMembership(ctx context.Context, req *gkehubpb.GetMembershipRequest, opts ...gax.CallOption) (*gkehubpb.Membership, error)
+}
+
+type GKEMembershipReconciler struct {
+	client GKEMembershipClient
+
+	Logger logr.Logger
+}
+
+func NewGKEClusterReconciler(
+	client GKEMembershipClient,
+	logger logr.Logger,
+) *GKEMembershipReconciler {
+	return &GKEMembershipReconciler{
+		client: client,
+		Logger: logger,
+	}
+}
+
+func (r *GKEMembershipReconciler) Reconcile(ctx context.Context, gcpCluster *capg.GCPCluster, oidcJwks []byte) (*gkehubpb.Membership, error) {
+	logger := r.Logger.WithValues("gkemembership", gcpCluster.Name)
+
+	membership := GenerateMembership(*gcpCluster, oidcJwks)
+	membershipExists, err := r.doesMembershipExist(ctx, membership.Name)
+	if err != nil {
+		logger.Error(err, "failed to check memberships existence")
+		return membership, err
+	}
+
+	if !membershipExists {
+		err = r.registerMembership(ctx, gcpCluster, membership)
+		if err != nil {
+			logger.Error(err, "failed to register cluster membership")
+			return membership, err
+		}
+
+		logger.Info(fmt.Sprintf("membership %s created", membership.Name))
+	}
+
+	return membership, nil
+}
+
+func GenerateMembership(cluster capg.GCPCluster, oidcJwks []byte) *gkehubpb.Membership {
+	externalId := uuid.New().String()
+
+	membershipId := GenerateMembershipId(cluster)
+	name := GenerateMembershipName(cluster)
+
+	workloadIdPool := GenerateWorkpoolId(cluster)
+	identityProvider := GenerateIdentityProvider(cluster, membershipId)
+
+	membership := &gkehubpb.Membership{
+		Name: name,
```

Comment:
The membership ID is actually passed in the request rather than the membership object. The external ID isn't necessary and I don't think we'll make use of it.
In any case we'll handle this in a separate PR